### PR TITLE
[DOCUMENT] Interfaces for the M3 Relay Protocol  (issue-#265)

### LINKED ISSUE #265: [DOCUMENT] Interfaces for the M3 Relay Protocol
## Objective

Code / document interfaces for the Relay Protocol to enable consensus and understanding around this piece of the M3 milestone.

## Origin Document

[M3 Milestone](https://github.com/pokt-network/pocket/blob/main/docs/roadmap/README.md#relay-protocol)

This is the first step into beginning the Relay Protocol development.

## Goals

- Clear interface(s) that show the lifecycle of the relay protocol
- Clear documentation supporting the above

## Deliverable

- Coded interface(s)
- Documentation support

## Non-goals / Non-deliverables

- Implementation of the interface

## General issue deliverables

- [ ] Update the appropriate CHANGELOG
- [ ] Update any relevant READMEs (local and/or global)
- [ ] Update any relevant global documentation & references
- [ ] If applicable, update the source code tree explanation
- [ ] If applicable, add or update a state, sequence or flowchart diagram using [mermaid](https://mermaid-js.github.io/mermaid/)

**Creator**: @andrewnguyen22

### PR DESCRIPTION
**Pull request by andrewnguyen22**
## Description

Code / document interfaces for the Relay Protocol to enable consensus and understanding around this piece of the M3 milestone.

[M3 Milestone](https://github.com/pokt-network/pocket/blob/main/docs/roadmap/README.md#relay-protocol)

This is the first step into beginning the Relay Protocol development.

- Clear interface(s) that show the lifecycle of the relay protocol
- Clear documentation supporting the above

## Issue

Fixes #265 

## Type of change

Please mark the relevant option(s):

- [ ] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Major breaking change
- [x] Documentation
- [ ] Other 

## List of changes

- Added Relay Protocol interfaces and diagrams
- 
## Testing

- [x] `make develop_test`
- [x] [LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md) w/ all of the steps outlined in the `README`

## Required Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have tested my changes using the available tooling
- [x] I have updated the corresponding CHANGELOG

### If Applicable Checklist
- [x] I have updated the corresponding README(s); local and/or global
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

### CODE REVIEW

File: utility/service/README.md [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+##### Alt Design
+
+NOTE: an alternative design is a 2 step, claim - proof lifecycle where the individual service nodes
+build a merkle sum index tree from all the relays, submits a root and subsequent merkle proof to the
+network.
+
+Pros: Can report volume metrics directly to the chain in a trustless fashion
+
+Cons: Large chain bloat, non-trivial compute requirement for creation of claim/proof transactions and trees, 
+non-trivial compute requirement to process claim / proofs during ApplyBlock()
```
Comment (Olshansk) [inline utility/service/README.md:159]:
```suggestion
## Alt Design

### Claim-Proof Lifecycle

An alternative design is a 2-step, claim-proof lifecycle where the individual service nodes
build a Merkle sum index tree from all the relays, submits a root and subsequent Merkle proof to the
network via a commit+reveal schema.

* **Pros**: Can report volume metrics directly to the chain in a trustless fashion
**Cons**: Large chain bloat, non-trivial compute requirement for creation of claim/proof transactions and trees, 
non-trivial compute requirement to process claim / proofs during ApplyBlock()

This algorithm is not yet documented anywhere, so the following links can act as a reference in the interim.

**Documentation:**
- Pocket docs: https://docs.pokt.network/home/v0/protocol/servicing#claim-proof-lifecycle
- Twitter Thread: https://twitter.com/o_rourke/status/1263847357122326530
- Plasma core Merkle Sum Tree: https://plasma-core.readthedocs.io/en/latest/specs/sum-tree.html

**Source code references:**
- Merkle: https://github.com/pokt-network/pocket-core/blob/staging/x/pocketcore/types/merkle.go
- Claim: https://github.com/pokt-network/pocket-core/blob/staging/x/pocketcore/keeper/claim.go
- Proof: https://github.com/pokt-network/pocket-core/blob/staging/x/pocketcore/keeper/proof.go
```

File: utility/service/README.md [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+```mermaid
+graph TD
+    A[Relay.Execute] -->|Lookup Relay.RelayChain URL|B
+    B[Chains.JSON -> RelayChainURL] -->|Execute Http Request| C
+    C[RelayChain] -->|Return Response| D
+    D[RelayResponse.Sign] --> |Send|E
+    E[Requester]
+```
+
+##### Alt Design
```
Comment (Olshansk) [inline utility/service/README.md:150]: Not a blocker and might be worth an offline discussion but just wanted to share an idea and am curious if:
1. This is something you're familiar with (or explored before) and know it's not going to work
2. You see any issues with the bat.

I've started looking into how Chia Network uses VDF (Verifiable Delay Functions) to have proof of space & time consensus (proof of committing storage to the network over some time). This work is based on Brem Cohen's Proof of Sequential work: https://eprint.iacr.org/2018/183.pdf.

In short, I'm not 100% if the math enables this yet, but if it does, this might be an alternative design as well, which could theoretically be simpler and more memory efficiency:

mermaid-diagram-2022-10-22-204928

**Review by Olshansk — CHANGES_REQUESTED** [review]
I haven't reviewed `utility/service/service.go` yet but left some comments on `utility/service/README.md` and figured I'd submit a first review.

Most NITS, a couple questions, and just one idea I wanted to get your thoughts on.

File: utility/service/README.md [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+The foundational lifecycle of the Relay Protocol is:
+
+1) Validate the inbound `Relay`
+2) Store / persist the `Relay`
+3) Execute the Relay against the `RelayChain`
+```mermaid
+sequenceDiagram
+	    autonumber
+	    actor App
+	    actor Client
```
[the remark this answers is not in the document]
Comment (andrewnguyen22) [reply utility/service/README.md:23]: I don't think participant is appropriate - in this case - the client is an `actor` within the `Relay Protocol`. Please present me an alternative if you don't find this satisfactory

File: utility/service/README.md [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+    C[RelayProtoBytes.Hash] -->|Add Hash to `Relay` and Persist| D
+    D[Relay.Persist] --> |Indexing Under Session|E
+    E[Key:SessionKey Val: Relays.AddNew]
+```
+
+### Execute the `Relay`
+
+Execute a submitted `Relay` against the `RelayChain` by a client after validation
+
+1) Retrieve the `RelayChain` url from chains.json file
```
[the remark this answers is not in the document]
Comment (andrewnguyen22) [reply utility/service/README.md:116]: Okay - totally can remove chains.json

I'm not sure what you're saying about on chain - but no - we don't store configuration on chain

File: utility/service/README.md [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+2) Execute http request with the `Relay Payload`
+3) Format and digitally sign the response
+4) Send back to client
+
+##### Wait for Session to end / secret key to be revealed
+
+It's important to note, the secret key isn't revealed by the network until the session is over
+to prevent volume based bias. The secret key is usually a pseudorandom selection using the block hash as a seed. 
+(See the session protocol)
```
[the remark this answers is not in the document]
Comment (andrewnguyen22) [reply utility/service/README.md:125]: It's a different key - but the same concept. 

Details blocked by [spike](https://github.com/pokt-network/pocket-network-protocol/issues/18)

File: utility/service/README.md [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+2) This function also signifies deleting the non-volume-applicable `Relays`
+
+### Report volume metric applicable relays to `Fisherman`
+
+1) All volume applicable relays need to be sent to the assigned trusted `Fisherman` for a proper verification of the volume completed.
+2) Send `volumeRelays` to `fishermanServiceURL` through http.
+
+```mermaid
+graph TD
+    A[Relay.Execute] -->|Lookup Relay.RelayChain URL|B
```
[the remark this answers is not in the document]
Comment (andrewnguyen22) [reply utility/service/README.md:143]: No - just the RelayChains you're registered to service

File: utility/service/README.md (hunk 2 above)
Comment (andrewnguyen22) [reply utility/service/README.md:150]: Sure - always open to new ideas. Chia's Proof of Time and Space was researched a bit during the consensus V1 spec'ing phase.

Would be interested to hear what you have in mind

Comment (Olshansk) [reply utility/service/README.md:150]: The underlying piece of proof of time & space are Verifiable Delay Functions (VDFs) which are based on Proof of Sequential work proofs (which are not functions).

After doing some more research and digging, I came to a similar solution to what you had with probabilistic hash collisions, but using Solana's proof of history[1]:

Screen Shot 2022-10-24 at 5 16 08 PM

I realize this diagram is getting a bit complicated so we wouldn't commit it as such, but hopefully we can use it for a discussion between ourselves:

```mermaid
sequenceDiagram
	    autonumber
        title Relay Volume Counter
	    actor C as Client
	    actor SN as Service Node
        participant IS as Internal Storage (Service Node)
        actor F as Fisherman
        C->>C: Generate SessionKey<br>(i.e. pseudo-random seed)
        C->>C: Initialize N=0
        C->>SN: SessionKey
	    loop Repeat N times - once per relay during session duration
            C->>+SN: SignedRelay
            SN->>SN: Validate & Process Relay
            SN->>SN: countHash=hashFunc(append(relayHash, countHash-1))<br>(using VDF)
            alt every M relays (e.g. 1000)
                SN->>IS: store(countHash)
            end
            SN->>-C: (countHash, SignedRelayResponse)
            C->>C: N += 1<br>update countHash
	    end
        C->>F: (SessionKey, countHash, N)
        SN->>F: (SessionKey, [countHashes], M)
        F->>F: assert(M == N)<br> Verify(SessionKey, [countHashes])
```

In some ways, Proof of history was built for verifiably proving increasing counters which gives me a gut sense that it may be a pointer in the right direction. 

To make the verification more optimal, it requires fishermen to be equipped with GPUs (see section 4.3 in [1]).

In relation to the hash collision approach, I think this ultimately just gives us a more deterministic workflow but I'm no longer sure it buys us any other benefits. 

[1] https://solana.com/solana-whitepaper.pdf

**Out of scope:** Eventually (not on v1 launch), we could also leverage the VDF to have a higher difficulty for RPC calls that "should cost more" (like eth archival calls), but I think that's outside the scope of this discussion but just a thought I had.

Comment (Olshansk) [reply utility/service/README.md:150]: If we do go with VDFs, we should probably use a golang wrapper around Rust libraries. 

Luckily, harmony already paid someone to do it: https://gitcoin.co/issue/25313

Comment (andrewnguyen22) [reply utility/service/README.md:150]: AFAIK this is a sub-optimal solution because it requires you to sequentially verify the hashes - am I misunderstanding?

Comment (Olshansk) [reply utility/service/README.md]: Will add a screenshot to the point I made earlier for more clarification.

> To make the verification more optimal, it requires fishermen to be equipped with GPUs (see section 4.3 in [1]).

Screen Shot 2022-10-25 at 7 31 55 AM

This will allow verification through parallelized hash verification which is guaranteed to have been executed sequentially.

The improvements I see are:
1. Deterministic vs probabilistic
2. Only involves verifying that the counts from the ServiceNode and Client are the same followed by the hash verification with the session context as the seed. Alternative, in order to avoid an attack where the ServiceNode starts pre-computing lots of hashes during the duration of a session, we need to open up the relay, verify the hash collision, check the signature of the client on each relay, and have some sort of deduping mechanism in place as well (which is avoidable by the sequential nature of the VDF).

Comment (Olshansk) [reply utility/service/README.md]: Don't want to block it, but do think it's worth tracking the discussion and decision.

I can move it to https://github.com/pokt-network/pocket/discussions if need be

Comment (Olshansk) [reply utility/service/README.md]: https://github.com/pokt-network/pocket/discussions/330

File: utility/service/service.go [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+
+type Relay interface {
+	RelayPayload
+	RelayMeta
+}
+
+type RelayPayload interface {
+	GetData() string               // the actual data string for the external chain
+	GetMethod() string             // the http CRUD method
+	GetPath() string               // the REST Path
```
Comment (Olshansk) [inline utility/service/service.go:16]: I think we should say `HTTP Path` unless it is actually rest.

Source:  https://htmx.org/essays/how-did-rest-come-to-mean-the-opposite-of-rest/?utm_source=pocket_mylist

Screen Shot 2022-11-02 at 7 59 28 PM

File: utility/service/PROTOCOLS.md [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+
+### Claim-Proof Lifecycle
+
+An alternative design is a 2-step, claim-proof lifecycle where the individual service nodes
+build a Merkle sum index tree from all the relays, submits a root and subsequent Merkle proof to the
+network via a commit+reveal schema.
+
+* **Pros**: Can report volume metrics directly to the chain in a trustless fashion
+**Cons**: Large chain bloat, non-trivial compute requirement for creation of claim/proof transactions and trees, 
+non-trivial compute requirement to process claim / proofs during ApplyBlock()
```
Comment (Olshansk) [inline utility/service/PROTOCOLS.md:164]:
```suggestion
* **Pros**: Can report volume metrics directly to the chain in a trustless fashion
* **Cons**: Large chain bloat, non-trivial compute requirement for creation of claim/proof transactions and trees, 
non-trivial compute requirement to process claim / proofs during ApplyBlock()
```

File: utility/service/service.go [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+
+	// persist relay object, indexing under session
+
+	return nil
+}
+
+// Execute a submitted relay by a client after validation
+func (r *relay) Execute() (RelayResponse, types.Error) {
+
+	// retrieve the RelayChain url from chains.json file
```
Comment (Olshansk) [inline utility/service/service.go:96]:
```suggestion
	// retrieve the RelayChain url from the servicer's local configuration file
```

**Review by Olshansk — CHANGES_REQUESTED** [review]

File: utility/service/PROTOCOLS.md [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+	    end
+```
+
+### Validate the inbound `Relay`
+
+A multi-step validation process to validate a submitted relay by a client before servicing
+
+1) Validate payload, look for empty or 'bad' request data
+2) Validate the metadata, look for empty or 'bad' metadata
+3) Ensure the `RelayChain` is supported locally (in the service node's configuration files)
```
[the remark this answers is not in the document]
Comment (andrewnguyen22) [reply utility/service/PROTOCOLS.md:70]: It's both technically, but definitely locally is a validation point

File: utility/service/service.go [hunk 11]
```diff
@@ -13,7 +13,7 @@ type Relay interface {
 type RelayPayload interface {
 	GetData() string               // the actual data string for the external chain
 	GetMethod() string             // the http CRUD method
-	GetPath() string               // the REST Path
+	GetHTTPPath() string           // the REST Path
```
Comment (Olshansk) [inline utility/service/service.go:16]:
```suggestion
	GetHTTPPath() string           // the HTTP Path
```

**Review by Olshansk — APPROVED** [review]

**Comment by Olshansk** [comment]
Approved with one minor remaining NIT.
